### ISSUE TITLE
"Error saving viewable of type numpy_image" on VSCode 1.85.1, this extension v3.0.12&v3.0.13

### ISSUE THREAD

**Issue by cleardusk** [elazarcoh/simply-view-image-for-python-debugging#68]
Thanks for the author, this extension helps me a lot 👍

I tried the code snippet below, `img_np` variable will raise "Error saving viewable of type numpy_image: AssertionError", but the `img_tensor` variable is fine and the image is shown correctly.

```python
import numpy as np  
import torch

img_np = np.clip(np.random.randn(1, 3, 256, 256) * 255, 0, 255).astype(np.uint8)   
img_tensor = torch.from_numpy(img_np).float() / 255.  
pass
```

**Comment by elazarcoh** [comment]
Thank you very much for the minimal example, it was very helpful to determine the problem.

The extension doesn't work with numpy images that are "channel first". Should be an easy fix, but I'm planning a major release in a very few weeks, so I'm preferring to have it there. If I see it's not going to be soon after all, I'll do it separately.

**Comment by cleardusk** [comment]
> Thank you very much for the minimal example, it was very helpful to determine the problem.
> 
> The extension doesn't work with numpy images that are "channel first". Should be an easy fix, but I'm planning a major release in a very few weeks, so I'm preferring to have it there. If I see it's not going to be soon after all, I'll do it separately.

Got it and thanks for your quick reply. So I guess `channel first` or `channel last` will be options in the configuration in the future?

And, very looking forward to your next major release! 😄

**Comment by elazarcoh** [comment]
I'm thinking of "guessing"whether it's first or last based on the shape. 
Probably something along:
3 or 4 dims & last dim > 4 & first or second dim (based on 3/4 dims) < 4.

What do you think? Does it capture most use cases?
If it does, then I can leave the rare use cases to be viewed using the "expression" feature to handle it more explicitly.

**Comment by cleardusk** [comment]
> I'm thinking of "guessing"whether it's first or last based on the shape. Probably something along: 3 or 4 dims & last dim > 4 & first or second dim (based on 3/4 dims) < 4.
> 
> What do you think? Does it capture most use cases? If it does, then I can leave the rare use cases to be viewed using the "expression" feature to handle it more explicitly.

I think, `ndarray` -> (Bx)HxWxC and `tensor` -> (Bx)xCxHxW cover most of the user cases and have become the consensus. What disturbs me (and perhaps other users) is the assertion error "Error saving viewable of type numpy_image: AssertionError." My initial reaction is that there might be a compatibility issue between the versions of VSCode and this extention, so I raised an issue here 😄

I prefer to let `channel_first` and `channel_last` be options to allow for more flexibility. If the user feeds a ndarray with shape `1x3x256x256`, the assertion error can add something like: `your ndarray variable seems to be channel first`, if the ndarray is with shape `1x32x256x256`, the error can be like `not support >4 channels`, etc.

Anyway, the above is just my personal perspective and does not represent other users, it is is more like a product decision-making challenge. I think users would likely accept both of them, as long as the description of this useful extention is clear~

**Comment by elazarcoh** [comment]
Actually the main issue is that it's been detected as an image rather than a tensor. There is a difference with how I handle images and tensors, when I'm saving it for display.

I totally agree about the error message, but it's not always clear what the issue is. I can understand it from the context, but it can also be the case that the user has a numpy array of 3 grayscale images. So I can't give a definite error message. I guess I can tell the user that the shape of the array is not supported, and suggest that it's a channel first.
I'll see how many complications this adds to the current logic and decide how to address it.

In any case, I'll make it a priority in the next release to handle those cases, and also will try to improve the error messages.

Thanks!

**Comment by cleardusk** [comment]
> Actually the main issue is that it's been detected as an image rather than a tensor. There is a difference with how I handle images and tensors, when I'm saving it for display.
> 
> I totally agree about the error message, but it's not always clear what the issue is. I can understand it from the context, but it can also be the case that the user has a numpy array of 3 grayscale images. So I can't give a definite error message. I guess I can tell the user that the shape of the array is not supported, and suggest that it's a channel first. I'll see how many complications this adds to the current logic and decide how to address it.
> 
> In any case, I'll make it a priority in the next release to handle those cases, and also will try to improve the error messages.
> 
> Thanks!

Looking forward to the next release 💖

**Comment by elazarcoh** [comment]
This is on pre-release! please check it out.
